### DISCUSSION TITLE
Comodulogram not showing

### DISCUSSION THREAD

**Discussion by sjg2203** [raphaelvallat/yasa#114] [Q&A]
While computing comodulogram, I can only get it to work for the mean vector length(MVL) but not for the modulation index (MI), which is the one that I need.
Would anyone have any idea why this isn't working?

I included my code and example of the comodulogram that doesn't work and the one that works.
```python
input_fname=easygui.fileopenbox(msg="EDF file")
raw=mne.io.read_raw_edf(input_fname,preload=True,verbose=False)
raw.resample(100)
sf=raw.info
raw.filter(0.1,45)
data=raw.get_data(units="uV")
sls=yasa.SleepStaging(raw,eeg_name='C3',eog_name='EOG R',emg_name='CHIN1')
hypno_pred=sls.predict()
hypno_pred=yasa.hypno_str_to_int(hypno_pred)
hypno_up=yasa.hypno_upsample_to_data(hypno_pred,sf_hypno=1/30,data=raw)
_,data_N2=yasa.sliding_window(hypno_up==2,sf,window=15)
data_N2.shape
f_pha=np.arange(0.375,4.25,0.25)
f_amp=np.arange(7.25,25.5,0.5)
from tensorpac import Pac
p=Pac(idpac=(2,0,0),f_pha=f_pha,f_amp=f_amp,verbose='WARNING')
xpac=p.filterfit(sf,data_N2)
p.comodulogram(xpac.mean(-1),title=str(p),vmin=0,plotas='contour',ncontours=25)
p.show()
```
```python
p=Pac(idpac=(2,0,0),f_pha=f_pha,f_amp=f_amp,verbose='WARNING')
```
No comodulogram showing for MI
![Comodulogram_MI](https://user-images.githubusercontent.com/58807831/206222579-45386c09-69db-430e-89e3-5236e076bb93.png)
but suddenly works when asking MVL
```python
p=Pac(idpac=(1,0,0),f_pha=f_pha,f_amp=f_amp,verbose='WARNING')
```
![Comodulogram_MVL](https://user-images.githubusercontent.com/58807831/206222511-67c282e3-c24c-4318-9ac3-6b805143734e.png)

Thanks

**Comment by raphaelvallat** [comment]
Hi,

It's hard to say without the original EDF data. Are you able to run [this notebook](https://github.com/raphaelvallat/yasa/blob/master/notebooks/12_SO-sigma_coupling.ipynb) in the notebooks/ folder of YASA? 

The MVL output looks kind of weird to m. I think there might be an issue in how you select the data. What does `data_N2` look like? Seems like it is currently a boolean?

**Reply by sjg2203** [reply]
I was able to run the notebook except for the **[example for one slow-wave](https://github.com/raphaelvallat/yasa/blob/b98a4c0029438162f3c96247c57d7de911a4525e/notebooks/12_SO-sigma_coupling.ipynb?short_path=1ea5862#L544)**, ran into this error. ThoughI could skip this part and get to the data-driven PAC part.
```python
>>ax1.plot(time,np.squeeze(sw._data_filt[:,idx_sw]),lw=2,color="k",label="Data (filtered)")

>>Traceback (most recent call last):
  File "C:\Users\serv_H\AppData\Local\Programs\Python\Python310\lib\site-packages\IPython\core\interactiveshell.py", line 3433, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-32-3127e985f09d>", line 1, in <module>
    ax1.plot(time,np.squeeze(sw._data_filt[:,idx_sw]),
  File "C:\Users\serv_H\AppData\Roaming\Python\Python310\site-packages\matplotlib\axes\_axes.py", line 1632, in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
  File "C:\Users\serv_H\AppData\Roaming\Python\Python310\site-packages\matplotlib\axes\_base.py", line 312, in __call__
    yield from self._plot_args(this, kwargs)
  File "C:\Users\serv_H\AppData\Roaming\Python\Python310\site-packages\matplotlib\axes\_base.py", line 498, in _plot_args
    raise ValueError(f"x and y must have same first dimension, but "
>>ValueError: x and y must have same first dimension, but have shapes (170,) and (26, 170)
```

As for data_N2, it's indeed an array of boolean 
![data_N2](https://user-images.githubusercontent.com/58807831/206852184-75ca9333-e1d5-422a-893f-d261ffbb870e.png)

**Reply by raphaelvallat** [reply]
Using `ax1.plot(time,np.squeeze(sw._data_filt[0,idx_sw]),lw=2,color="k",label="Data (filtered)")` should fix your issue — in other words focusing on a single channel at a time for plotting.

To calculate the PAC, you must pass the raw data and not a boolean array. What is your `sf` and `data_N2.shape`?

**Reply by sjg2203** [reply]
I define `sf` myself as I resampled its frequency to 100Hz
```python
sf=100
```
```python
data_N2.shape
Output: (2397,1500)
```

**Reply by raphaelvallat** [reply]
What happens if you use the raw data instead of bool for `data_N2`, and also perhaps limit the number of epochs that you are using in the computation? e.g. `data_N2[:100, :]` to select the first 100 epochs

**Reply by sjg2203** [reply]
I amended `data_N2` so that it uses the raw data instead, and also cut the 100 first epochs.
```python
_,data_N2=yasa.sliding_window(data,sf,window=15)
data_N2=data_N2[:100,:]
data_N2.shape
Output: (100,26,1500)
```
However, it returned this error:
```python
>>xpac=p.filterfit(sf,data_N2)

>>Traceback (most recent call last):
  File "C:\Users\serv_H\AppData\Local\Programs\Python\Python310\lib\site-packages\IPython\core\interactiveshell.py", line 3433, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-83-fc1fe6e5d6a5>", line 1, in <module>
    xpac=p.filterfit(sf,data_N2)
  File "C:\Users\serv_H\AppData\Local\Programs\Python\Python310\lib\site-packages\tensorpac\pac.py", line 478, in filterfit
    pha = self.filter(sf, x_pha, 'phase', **kw)
  File "C:\Users\serv_H\AppData\Local\Programs\Python\Python310\lib\site-packages\tensorpac\pac.py", line 78, in filter
    assert x.ndim == 2, ("x should be a 2d array like (n_epochs, n_times)")
>>AssertionError: x should be a 2d array like (n_epochs, n_times)
```

**Reply by raphaelvallat** [reply]
In your latest example `data_N2` is a 3D array of shape (100, 26, 1500). Where does the 26 comes from? Is that the number of channels? You can only calculate the Pac on a single channel at a time. Please use something like `np.squeeze(data_N2[:, 0, :])`

**Reply by sjg2203** [reply]
26 is indeed the number of channels in my edf file. 
I used what you suggested and it now generates a much better plot, though I am not entirely sure why there is some merging happening here.
```python
>>data_N2=np.squeeze(data_N2[:,0,:])
>>data_N2.shape
Output: (2030, 1500)
```
![Comodulogram_MI2](https://user-images.githubusercontent.com/58807831/207463121-1d24cd06-3803-4657-b7b3-80d3475eb9de.png)

**Reply by raphaelvallat** [reply]
What do you mean by merging?

**Reply by sjg2203** [reply]
The three main blobs are somehow meeting, I mean, normally we do only have two of those which are quite distinct but here, we can see some gradient going from one another
